// uu5imagingg01/dist-esm/imaging/carousel.js
//@@viewOn:imports
import UU5 from "uu5g04";
import Cfg from "../config/config.js";
import CarouselBody from "./internal/carousel-body.js";
//@@viewOff:imports

let Carousel = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Cfg.TAG + "Carousel",
    nestingLevelList: UU5.Environment.getNestingLevelList("BigBoxCollection", "box"),
    classNames: {
      main: props => {
        let classes = [Cfg.CSS + "carousel"];

        if (props.height) {
          classes.push(UU5.Common.Css.css`
            && {
              height: ${UU5.Common.Tools.fillUnit(props.height)};
            }
          `);
        }

        return classes.join(" ");
      }
    },
    opt: {
      hoc: true,
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    data: UU5.PropTypes.any,
    authenticate: UU5.PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      data: null,
      authenticate: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _renderResizeBody({ width }, orphan) {
    if (width === 0) {
      // fix for Resize rendering without dimensions first
      return null;
    } else {
      return this._renderBody(width, orphan);
    }
  },

  _renderBody(width = this.props.width, orphan) {
    let data = this.props.data || [];

    if (this.props.authenticate) {
      return (
        <UU5.Common.SessionContext.Consumer>
          {session => (
            <CarouselBody
              {...this.props}
              {...this._getMainPropsToPass(orphan)}
              data={data}
              width={width}
              session={session ? session.session : undefined}
            />
          )}
        </UU5.Common.SessionContext.Consumer>
      );
    } else {
      return (
        <CarouselBody
          {...this.props}
          {...this._getMainPropsToPass(orphan)}
          data={data}
          width={width}
          parent={null}
          nestingLevel={undefined}
        />
      );
    }
  },

  _getMainPropsToPass(orphan) {
    let result = this.getMainPropsToPass();
    if (orphan) {
      delete result.parent;
      delete result.nestingLevel;
    }
    return result;
  },

  _getChildren(orphan) {
    if (this.props.width) {
      return this._renderBody(undefined, orphan);
    } else {
      return (
        <UU5.Bricks.Resize {...this._getMainPropsToPass(orphan)} nestingLevel={undefined} ref_={null} parent={null}>
          {size => this._renderResizeBody(size, orphan)}
        </UU5.Bricks.Resize>
      );
    }
  },

  _getLsiLabel() {
    return (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Prezentace obrázků",
          en: "Show slideshow"
        }}
      />
    );
  },
  //@@viewOff:private
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() ? (
      this._getChildren()
    ) : (
      <UU5.Bricks.LinkModal component={this._getChildren(true)}>{this._getLsiLabel()}</UU5.Bricks.LinkModal>
    );
  }
  //@@viewOff:render
});

Carousel = UU5.Common.withVisibilityCheck(Carousel);

export { Carousel };
export default Carousel;
